consensus/src/lib.rs

```diff
@@ -78,165 +70,318 @@ pub trait PolkadotClient: Clone + 'static {
 	) -> ClientResult<Option<Vec<u8>>>;
 }
 
-/// Finalize the given block in the Parachain.
-fn finalize_block<T, Block, B>(client: &T, hash: Block::Hash) -> ClientResult<bool>
+/// Follow the finalized head of the given parachain.
+///
+/// For every finalized block of the relay chain,
+async fn follow_finalized_head<P, Block, B, R>(
+	para_id: ParaId,
+	parachain: Arc<P>,
+	relay_chain: R,
+) -> ClientResult<()>
 where
 	Block: BlockT,
-	T: Finalizer<Block, B> + UsageProvider<Block>,
+	P: Finalizer<Block, B> + UsageProvider<Block>,
+	R: RelaychainClient,
 	B: Backend<Block>,
 {
-	// don't finalize the same block multiple times.
-	if client.usage_info().chain.finalized_hash != hash {
-		match client.finalize_block(BlockId::hash(hash), None, true) {
-			Ok(()) => Ok(true),
-			Err(e) => match e {
-				ClientError::UnknownBlock(_) => Ok(false),
-				_ => Err(e),
-			},
+	let mut finalized_heads = relay_chain.finalized_heads(para_id)?;
+
+	loop {
+		let finalized_head = if let Some(h) = finalized_heads.next().await {
+			h
+		} else {
+			tracing::debug!(target: "cumulus-consensus", "Stopping following finalized head.");
+			return Ok(());
+		};
+
+		let header = match Block::Header::decode(&mut &finalized_head[..]) {
+			Ok(header) => header,
+			Err(err) => {
+				tracing::warn!(
+					target: "cumulus-consensus",
+					error = ?err,
+					"Could not decode parachain header while following finalized heads.",
+				);
+				continue;
+			}
+		};
+
+		let hash = header.hash();
+
+		// don't finalize the same block multiple times.
+		if parachain.usage_info().chain.finalized_hash != hash {
+			if let Err(e) = parachain.finalize_block(BlockId::hash(hash), None, true) {
+				match e {
+					ClientError::UnknownBlock(_) => tracing::debug!(
```

Comment:
I suppose here there is a similar error to the one that the PR fixes. If the relay chain informs us about a finalized header before we've imported that header, then we won't finalize it, even if we import that same header immediately afterwards.

We will correctly finalize probably soon after, when we get a finality notification corresponding to a header we have already imported.

Reply:
That is right, but also done intentionally by me. I do not expect that we are that much behind the tip that we can not finalize the block, because we did not have this block yet. If that is happening, we probably have other problems.

Reply:
Maybe should be filed for the "wishlist" but yeah I agree. Even the slight lag in finality should mean that blocks are well propagated before finality.